### PR TITLE
perf(memory-hybrid): workflow patterns telemetry and duplicate detection

### LINKED ISSUE #1415: Epic: Workflow patterns, telemetry IO, and duplicate detection perf (supersedes #1399, #1400, #1401)
## Combined coder briefing

_Structured extracts from each superseded issue (plus bold **Description** / **Impact** blocks when issues do not use `## Summary`). Verbatim originals remain under **Reference**._

### #1399: Auto-skills: WorkflowStore.getPatterns is O(N×C×L²) per call
**Summary / description**

`WorkflowStore.getPatterns()` loads every row in `workflow_traces` and clusters them in JavaScript using `sequenceSimilarity` (Levenshtein over tool sequences). Complexity is `O(N × C × L²)` per call, where `N` is total traces, `C` is the number of growing clusters, and `L` is sequence length. This runs on every crystallization cycle (and on every CLI `workflows` invocation).

**Impact**

- For a workspace with 5k traces and ~30 emerging clusters at avg sequence length 6, that is roughly `5000 × 30 × 36 ≈ 5.4M` cell operations per call. Empirically 2–8 seconds on cold caches, blocks the main thread.
- Memory pressure: parsed clusters retain all goals/outcomes/durations arrays for the lifetime of the call.
- This is on the critical path of `runCycle` (called from cron and the `memory_crystallize` agent tool).

**Evidence / code**

```ts
// extensions/memory-hybrid/backends/workflow-store.ts — getPatterns()
const allRows = this.liveDb
  .prepare("SELECT goal, tool_sequence, outcome, duration_ms FROM workflow_traces ORDER BY created_at DESC")
  .all() as { ... }[];

const clusters: { representative: string[]; goals: string[]; outcomes: string[]; durations: number[] }[] = [];

for (const row of allRows) {
  let seq: string[];
  try { seq = JSON.parse(row.tool_sequence) as string[]; } catch { continue; }
  let found = false;
  for (const cluster of clusters) {
    if (sequenceSimilarity(seq, cluster.representative) >= threshold) {
      cluster.goals.push(row.goal);
      // ...
      found = true;
      break;
    }
  }
  if (!found) clusters.push({ representative: seq, ... });
}
```

`sequenceDistance` (Levenshtein) allocates an `(m+1)×(n+1)` matrix per pair.

**Suggested fix**

1. **Bound the input.** Add a `limit` parameter to the SELECT (e.g. 1000 newest traces); the older tail rarely changes cluster membership.
2. **Index by argsHash or sequence shape.** Group rows by `args_hash` first (already indexed) and only run Levenshtein within groups that share the hash; cross-group similarity below threshold is unlikely.
3. **Persist clusters incrementally.** Maintain a `workflow_clusters` table updated on `record()`; `getPatterns()` becomes a simple `SELECT` + small per-cluster aggregation.
4. **Memoise** the JS clustering result by `(cutoff_timestamp, total_count)` until a new trace is inserted.

Pick (1)+(2) as a quick fix, plan (3) as a longer-term refactor.

**Acceptance criteria**

- [ ] `getPatterns()` runtime is sub-second for 10k traces in a representative benchmark.
- [ ] Crystallization cycle test asserts deterministic patterns across runs.
- [ ] No regression in `memory_workflows` tool output for existing fixtures.

**Files / areas**

- `extensions/memory-hybrid/backends/workflow-store.ts`
- `extensions/memory-hybrid/services/pattern-detector.ts`
- (optional) new `extensions/memory-hybrid/backends/workflow-cluster-store.ts`

---

### #1400: Auto-skills: recordGeneratedSkillTelemetry re-scans every activation on each insert
**Summary / description**

Every call to `recordGeneratedSkillTelemetry` ends with `refreshGeneratedSkillLifecycleState`, which executes `skillTelemetryEntries(db, skillName)` — a `SELECT * FROM generated_skill_telemetry WHERE skill_name = ? ORDER BY created_at DESC` with **no LIMIT** — and re-summarises all activations for the skill. For a popular skill this turns activation insertion into O(N) per insert and the cumulative cost over a day into O(N²).

`buildGeneratedSkillTelemetryReport` further compounds this by calling `listGeneratedSkillProcedures` twice (once to refresh state for each procedure, then again to map rows for the response).

**Impact**

- A skill with thousands of activations becomes increasingly expensive to record telemetry against, eventually noticeable as latency on `skills record`.
- `buildGeneratedSkillTelemetryReport` doubles the cost of listing.

**Evidence / code**

```ts
// extensions/memory-hybrid/backends/facts-db/generated-skills.ts
export function recordGeneratedSkillTelemetry(...) {
  // ... INSERT ...
  refreshGeneratedSkillLifecycleState(db, ..., policy, now);  // re-summarises ALL rows
  return mapGeneratedSkillTelemetryRow(...);
}

function skillTelemetryEntries(db: DatabaseSync, skillName: string): GeneratedSkillTelemetryEntry[] {
  return (
    db.prepare(`SELECT * FROM generated_skill_telemetry WHERE skill_name = ? ORDER BY created_at DESC`)
      .all(skillName) as GeneratedSkillTelemetryRow[]
  ).map(mapGeneratedSkillTelemetryRow);
}

export function buildGeneratedSkillTelemetryReport(...) {
  const procedures = listGeneratedSkillProcedures(db).filter(...);  // first pass
  for (const proc of procedures) {
    if (proc.skillPath) refreshGeneratedSkillLifecycleState(db, basename(proc.skillPath), policy, now);
  }
  const rows = listGeneratedSkillProcedures(db)                     // second pass!
    .filter(...)
    .map(...);
  // ...
}
```

**Suggested fix**

1. **Maintain rollup counters** on the procedure row (or in a sidecar `generated_skill_metrics` table): `selected_count`, `near_miss_count`, `false_positive_count`, `last_used_at`, ... Update them on every telemetry insert via a single SQL `UPDATE`.
2. **Compute flags from rollups**, not from full row scans, except on explicit `refresh --full`.
3. **Single-pass `buildGeneratedSkillTelemetryReport`** — call `listGeneratedSkillProcedures` once, refresh while iterating, build the row inline.

Net effect: telemetry insert becomes O(1); report becomes O(distinct skills) instead of O(activations).

**Acceptance criteria**

- [ ] `recordGeneratedSkillTelemetry` does not perform an unbounded `SELECT *` per insert.
- [ ] `buildGeneratedSkillTelemetryReport` lists procedures exactly once per invocation.
- [ ] Benchmark / micro-test demonstrates O(1) insert cost.
- [ ] Existing telemetry tests still pass (rollups produce same numbers as the current scan).

**Files / areas**

- `extensions/memory-hybrid/backends/facts-db/generated-skills.ts`
- `extensions/memory-hybrid/backends/facts-db/facts-db-layer2.ts`
- `extensions/memory-hybrid/tests/generated-skill-telemetry.test.ts`

---

### #1401: Auto-skills: isDuplicateSkill reads every SKILL.md from disk per promotion attempt
**Summary / description**

`isDuplicateSkill` (procedure pipeline duplicate-detection gate) walks every existing skill directory and reads each `SKILL.md` on **every promotion attempt**. With 100 skills that is ~100 file reads per promotion; with 1000+ skills the gate becomes the slowest part of `generate-auto-skills`.

**Impact**

- Slow `generate-auto-skills --apply` runs.
- Read amplification on CI runners.
- Failure if the workspace has a slow filesystem (NFS-mounted skills dirs, sandboxed runners with slow `readdir`).

**Evidence / code**

```ts
// extensions/memory-hybrid/services/procedure-promotion-policy.ts
function isDuplicateSkill(slug, task, skillsAutoPath, extraDirs = [], inRunCandidates = []): boolean {
  // ...
  for (const dir of dirs) {
    if (!existsSync(dir)) continue;
    for (const entry of safeReadDir(dir)) {
      const skillPath = join(dir, entry, "SKILL.md");
      if (!existsSync(skillPath)) continue;
      if (entry === slug) return true;
      const content = safeReadFile(skillPath).toLowerCase();    // ← reads every file
      if (content.includes(`name: ${slug}`)) return true;
      const taskContent = extractTaskContentFromSkill(content);
      const contentWords = significantWords(taskContent);
      // ...
    }
  }
  return false;
}
```

This is called once per procedure during `evaluateProcedureForPromotion`. With `MAX_SKILLS_PER_RUN = 10` and 200 existing skills, a single `--apply` batch reads 2000 files (more if `extraDirs` is populated).

**Suggested fix**

1. **Cache** the digested keyword set per `(absolute_skill_path, mtime)` in a module-level `Map`. Subsequent calls within the same process reuse the digest unless mtime changes.
2. **Index file** in each skills root (`.openclaw-skills-index.json`) that stores `{ slug → digest }`; written by `writeDraftSkill` and `writeSkillToDisk` whenever a skill is created. `isDuplicateSkill` reads only the index.
3. **Persist digest in DB** keyed by `skillPath`; read from DB instead of disk.

Pair with #1372 (rebase slug collision) — the same index file would also detect inconsistencies.

**Acceptance criteria**

- [ ] `isDuplicateSkill` does not read every `SKILL.md` from disk for each promotion attempt.
- [ ] Cache invalidation on mtime change or explicit `--no-cache` flag.
- [ ] Benchmark / test showing scaling improvement with N skills.

**Files / areas**

- `extensions/memory-hybrid/services/procedure-promotion-policy.ts`
- `extensions/memory-hybrid/services/procedure-skill-generator.ts`

---

## Merged acceptance checklist (deduped)

- [ ] `getPatterns()` runtime is sub-second for 10k traces in a representative benchmark.
- [ ] Crystallization cycle test asserts deterministic patterns across runs.
- [ ] No regression in `memory_workflows` tool output for existing fixtures.
- [ ] `recordGeneratedSkillTelemetry` does not perform an unbounded `SELECT *` per insert.
- [ ] `buildGeneratedSkillTelemetryReport` lists procedures exactly once per invocation.
- [ ] Benchmark / micro-test demonstrates O(1) insert cost.
- [ ] Existing telemetry tests still pass (rollups produce same numbers as the current scan).
- [ ] `isDuplicateSkill` does not read every `SKILL.md` from disk for each promotion attempt.
- [ ] Cache invalidation on mtime change or explicit `--no-cache` flag.
- [ ] Benchmark / test showing scaling improvement with N skills.

## Quick links (superseded issues)

- https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1399
- https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1400
- https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1401

## Supersedes
- #1399 — `WorkflowStore.getPatterns` complexity / full table scan.
- #1400 — Telemetry full rescan per insert + double list in report.
- #1401 — `isDuplicateSkill` reads all SKILL.md files per promotion.

---

## Reference: merged detail from superseded issues

_Verbatim closed-issue bodies (expand any row):_

<details>
<summary><strong>#1399: Auto-skills: WorkflowStore.getPatterns is O(N×C×L²) per call</strong></summary>

## Summary

`WorkflowStore.getPatterns()` loads every row in `workflow_traces` and clusters them in JavaScript using `sequenceSimilarity` (Levenshtein over tool sequences). Complexity is `O(N × C × L²)` per call, where `N` is total traces, `C` is the number of growing clusters, and `L` is sequence length. This runs on every crystallization cycle (and on every CLI `workflows` invocation).

## Evidence

```ts
// extensions/memory-hybrid/backends/workflow-store.ts — getPatterns()
const allRows = this.liveDb
  .prepare("SELECT goal, tool_sequence, outcome, duration_ms FROM workflow_traces ORDER BY created_at DESC")
  .all() as { ... }[];

const clusters: { representative: string[]; goals: string[]; outcomes: string[]; durations: number[] }[] = [];

for (const row of allRows) {
  let seq: string[];
  try { seq = JSON.parse(row.tool_sequence) as string[]; } catch { continue; }
  let found = false;
  for (const cluster of clusters) {
    if (sequenceSimilarity(seq, cluster.representative) >= threshold) {
      cluster.goals.push(row.goal);
      // ...
      found = true;
      break;
    }
  }
  if (!found) clusters.push({ representative: seq, ... });
}
```

`sequenceDistance` (Levenshtein) allocates an `(m+1)×(n+1)` matrix per pair.

## Impact

- For a workspace with 5k traces and ~30 emerging clusters at avg sequence length 6, that is roughly `5000 × 30 × 36 ≈ 5.4M` cell operations per call. Empirically 2–8 seconds on cold caches, blocks the main thread.
- Memory pressure: parsed clusters retain all goals/outcomes/durations arrays for the lifetime of the call.
- This is on the critical path of `runCycle` (called from cron and the `memory_crystallize` agent tool).

## Suggested fix

1. **Bound the input.** Add a `limit` parameter to the SELECT (e.g. 1000 newest traces); the older tail rarely changes cluster membership.
2. **Index by argsHash or sequence shape.** Group rows by `args_hash` first (already indexed) and only run Levenshtein within groups that share the hash; cross-group similarity below threshold is unlikely.
3. **Persist clusters incrementally.** Maintain a `workflow_clusters` table updated on `record()`; `getPatterns()` becomes a simple `SELECT` + small per-cluster aggregation.
4. **Memoise** the JS clustering result by `(cutoff_timestamp, total_count)` until a new trace is inserted.

Pick (1)+(2) as a quick fix, plan (3) as a longer-term refactor.

## Acceptance criteria

- [ ] `getPatterns()` runtime is sub-second for 10k traces in a representative benchmark.
- [ ] Crystallization cycle test asserts deterministic patterns across runs.
- [ ] No regression in `memory_workflows` tool output for existing fixtures.

## Files

- `extensions/memory-hybrid/backends/workflow-store.ts`
- `extensions/memory-hybrid/services/pattern-detector.ts`
- (optional) new `extensions/memory-hybrid/backends/workflow-cluster-store.ts`

</details>

<details>
<summary><strong>#1400: Auto-skills: recordGeneratedSkillTelemetry re-scans every activation on each insert</strong></summary>

## Summary

Every call to `recordGeneratedSkillTelemetry` ends with `refreshGeneratedSkillLifecycleState`, which executes `skillTelemetryEntries(db, skillName)` — a `SELECT * FROM generated_skill_telemetry WHERE skill_name = ? ORDER BY created_at DESC` with **no LIMIT** — and re-summarises all activations for the skill. For a popular skill this turns activation insertion into O(N) per insert and the cumulative cost over a day into O(N²).

`buildGeneratedSkillTelemetryReport` further compounds this by calling `listGeneratedSkillProcedures` twice (once to refresh state for each procedure, then again to map rows for the response).

## Evidence

```ts
// extensions/memory-hybrid/backends/facts-db/generated-skills.ts
export function recordGeneratedSkillTelemetry(...) {
  // ... INSERT ...
  refreshGeneratedSkillLifecycleState(db, ..., policy, now);  // re-summarises ALL rows
  return mapGeneratedSkillTelemetryRow(...);
}

function skillTelemetryEntries(db: DatabaseSync, skillName: string): GeneratedSkillTelemetryEntry[] {
  return (
    db.prepare(`SELECT * FROM generated_skill_telemetry WHERE skill_name = ? ORDER BY created_at DESC`)
      .all(skillName) as GeneratedSkillTelemetryRow[]
  ).map(mapGeneratedSkillTelemetryRow);
}

export function buildGeneratedSkillTelemetryReport(...) {
  const procedures = listGeneratedSkillProcedures(db).filter(...);  // first pass
  for (const proc of procedures) {
    if (proc.skillPath) refreshGeneratedSkillLifecycleState(db, basename(proc.skillPath), policy, now);
  }
  const rows = listGeneratedSkillProcedures(db)                     // second pass!
    .filter(...)
    .map(...);
  // ...
}
```

## Impact

- A skill with thousands of activations becomes increasingly expensive to record telemetry against, eventually noticeable as latency on `skills record`.
- `buildGeneratedSkillTelemetryReport` doubles the cost of listing.

## Suggested fix

1. **Maintain rollup counters** on the procedure row (or in a sidecar `generated_skill_metrics` table): `selected_count`, `near_miss_count`, `false_positive_count`, `last_used_at`, ... Update them on every telemetry insert via a single SQL `UPDATE`.
2. **Compute flags from rollups**, not from full row scans, except on explicit `refresh --full`.
3. **Single-pass `buildGeneratedSkillTelemetryReport`** — call `listGeneratedSkillProcedures` once, refresh while iterating, build the row inline.

Net effect: telemetry insert becomes O(1); report becomes O(distinct skills) instead of O(activations).

## Acceptance criteria

- [ ] `recordGeneratedSkillTelemetry` does not perform an unbounded `SELECT *` per insert.
- [ ] `buildGeneratedSkillTelemetryReport` lists procedures exactly once per invocation.
- [ ] Benchmark / micro-test demonstrates O(1) insert cost.
- [ ] Existing telemetry tests still pass (rollups produce same numbers as the current scan).

## Files

- `extensions/memory-hybrid/backends/facts-db/generated-skills.ts`
- `extensions/memory-hybrid/backends/facts-db/facts-db-layer2.ts`
- `extensions/memory-hybrid/tests/generated-skill-telemetry.test.ts`

</details>

<details>
<summary><strong>#1401: Auto-skills: isDuplicateSkill reads every SKILL.md from disk per promotion attempt</strong></summary>

## Summary

`isDuplicateSkill` (procedure pipeline duplicate-detection gate) walks every existing skill directory and reads each `SKILL.md` on **every promotion attempt**. With 100 skills that is ~100 file reads per promotion; with 1000+ skills the gate becomes the slowest part of `generate-auto-skills`.

## Evidence

```ts
// extensions/memory-hybrid/services/procedure-promotion-policy.ts
function isDuplicateSkill(slug, task, skillsAutoPath, extraDirs = [], inRunCandidates = []): boolean {
  // ...
  for (const dir of dirs) {
    if (!existsSync(dir)) continue;
    for (const entry of safeReadDir(dir)) {
      const skillPath = join(dir, entry, "SKILL.md");
      if (!existsSync(skillPath)) continue;
      if (entry === slug) return true;
      const content = safeReadFile(skillPath).toLowerCase();    // ← reads every file
      if (content.includes(`name: ${slug}`)) return true;
      const taskContent = extractTaskContentFromSkill(content);
      const contentWords = significantWords(taskContent);
      // ...
    }
  }
  return false;
}
```

This is called once per procedure during `evaluateProcedureForPromotion`. With `MAX_SKILLS_PER_RUN = 10` and 200 existing skills, a single `--apply` batch reads 2000 files (more if `extraDirs` is populated).

## Impact

- Slow `generate-auto-skills --apply` runs.
- Read amplification on CI runners.
- Failure if the workspace has a slow filesystem (NFS-mounted skills dirs, sandboxed runners with slow `readdir`).

## Suggested fix

1. **Cache** the digested keyword set per `(absolute_skill_path, mtime)` in a module-level `Map`. Subsequent calls within the same process reuse the digest unless mtime changes.
2. **Index file** in each skills root (`.openclaw-skills-index.json`) that stores `{ slug → digest }`; written by `writeDraftSkill` and `writeSkillToDisk` whenever a skill is created. `isDuplicateSkill` reads only the index.
3. **Persist digest in DB** keyed by `skillPath`; read from DB instead of disk.

Pair with #1372 (rebase slug collision) — the same index file would also detect inconsistencies.

## Acceptance criteria

- [ ] `isDuplicateSkill` does not read every `SKILL.md` from disk for each promotion attempt.
- [ ] Cache invalidation on mtime change or explicit `--no-cache` flag.
- [ ] Benchmark / test showing scaling improvement with N skills.

## Files

- `extensions/memory-hybrid/services/procedure-promotion-policy.ts`
- `extensions/memory-hybrid/services/procedure-skill-generator.ts`

</details>

### PR DESCRIPTION
**Pull request by markus-lassfolk**
Closes https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1415

## #1399 WorkflowStore.getPatterns
- Bounded `SELECT` (default `traceSampleLimit: 2000`, overridable).
- Cluster Levenshtein **within** `args_hash` buckets to cut cross-comparisons.
- **Memoization** on table `(COUNT, MAX(created_at))` fingerprint + options; invalidated on `record()` / `prune()`.
- Pattern detector passes `traceSampleLimit: 6000` for headroom.

## #1400 Telemetry IO
- New `gst_*` rollup columns on `procedures` (migration + one-time backfill when columns first appear).
- `recordGeneratedSkillTelemetry`: **O(1)** delta `UPDATE` after insert; lifecycle refresh uses rollups + a single **COUNT** for last-7d activations (no full `SELECT *` scan).
- `buildGeneratedSkillTelemetryReport`: **single** `listGeneratedSkillProcedures`; recent rows use explicit column list + `LIMIT`.
- `rebuildGeneratedSkillTelemetryRollupsForProcedure` for correction paths (`markGeneratedSkillTelemetryFalsePositive`).

## #1401 isDuplicateSkill
- **mtime cache** per absolute `SKILL.md` path (re-read when mtime changes).
- `--bypass-skill-duplicate-cache` on `generate-auto-skills` → `bypassDuplicateSkillCache` on promotion options.
- `clearProcedurePromotionDuplicateSkillCache()` for tests.

## Tests
- Memo determinism; sampled `getPatterns` perf smoke; rollup vs rebuild parity.

Commit author: **Markus Lassfolk** (noreply GitHub email).

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Touches SQLite schema/migrations and changes how generated-skill telemetry metrics are computed (rollup deltas + backfills), which could skew lifecycle decisions/reports if counters drift. Also adds memoization/sampling to workflow pattern detection, which is performance-driven but could change which patterns surface under some option combinations.
> 
> **Overview**
> **Performance-focused updates to memory-hybrid telemetry and skill generation.** Generated-skill telemetry reporting/lifecycle refresh is reworked to use new per-`procedure` `gst_*` rollup columns that are incrementally updated on telemetry insert, with an idempotent migration/backfill and a rebuild helper used after correction edits; reports now fetch only a limited set of recent activation rows and ignore negative “savings” values.
> 
> **Workflow pattern detection is bounded and cached.** `WorkflowStore.getPatterns` now samples a limited number of newest traces, memoizes results (with defensive copies) and invalidates on `record`/`prune`; `PatternDetector` opts into a higher sample limit.
> 
> **Duplicate-skill detection is cached + controllable.** `SKILL.md` reads are cached by mtime via a small LRU, with a new `--bypass-skill-duplicate-cache` flag plumbed through the CLI and skill generator to force re-reads; tests added for rollup parity, memoization behavior, sampling performance, and clustering edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 47551faf229829be3d62c131ac99aa6796aba6d8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: extensions/memory-hybrid/backends/workflow-store.ts [hunk 1]
```diff
@@ -352,47 +367,92 @@ export class WorkflowStore extends BaseSqliteStore {
...
+      const minRate = options?.minSuccessRate ?? 0;
+      const resultLimit = options?.limit ?? 20;
+      const meta = this.liveDb.prepare("SELECT COUNT(*) as c, MAX(created_at) as m FROM workflow_traces").get() as {
+        c: number;
+        m: string | null;
+      };
+      const memo = this.workflowPatternsMemo;
+      if (
+        memo &&
+        memo.rowCount === meta.c &&
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/workflow-store.ts]: Fixed in becd6a2704a3 — getPatterns memo validation now uses an in-memory mutation revision invalidated on record/prune, removing the per-call COUNT/MAX table scan.

File: extensions/memory-hybrid/backends/workflow-store.ts [hunk 2]
```diff
@@ -352,47 +367,92 @@ export class WorkflowStore extends BaseSqliteStore {
...
-        try {
-          seq = JSON.parse(row.tool_sequence) as string[];
-        } catch {
-          continue;
-        }
+      for (const group of byHash.values()) {
+        for (const row of group) {
+          let seq: string[];
+          try {
+            seq = JSON.parse(row.tool_sequence) as string[];
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/workflow-store.ts]: Fixed in becd6a2704a3 / prior branch commits — clustering is isolated per args_hash bucket with per-bucket clusters concatenated into allClusters.

File: extensions/memory-hybrid/backends/migrations/facts-migrations.ts [hunk 3]
```diff
@@ -427,6 +427,68 @@ function migrateGeneratedSkillTelemetryTable(db: DatabaseSync): void {
...
+/** O(1) telemetry rollups on procedures (#1415 / #1400). */
+function migrateGeneratedSkillTelemetryRollupColumns(db: DatabaseSync): void {
+  const cols0 = db.prepare("PRAGMA table_info(procedures)").all() as Array<{ name: string }>;
+  const hadGst = new Set(cols0.map((c) => c.name)).has("gst_sel_total");
+  const colNames = new Set(cols0.map((c) => c.name));
+  const add = (name: string, ddl: string) => {
+    if (!colNames.has(name)) {
+      db.exec(ddl);
+      colNames.add(name);
+    }
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/migrations/facts-migrations.ts:445]: Fixed in becd6a2704a3 — the rollup migration now tracks whether any rollup column was newly added and backfills in that case, covering partial/crashed migrations.

File: extensions/memory-hybrid/backends/facts-db/generated-skills.ts [hunk 4]
```diff
@@ -504,22 +769,17 @@ export function buildGeneratedSkillTelemetryReport(
...
-      if (!options?.skillName) return true;
-      return proc.skillPath != null && basename(proc.skillPath) === options.skillName;
-    })
+  const rows = procedures
     .map((proc) => {
       const skillName = basename(proc.skillPath ?? proc.taskPattern);
-      const activations = skillTelemetryEntries(db, skillName);
-      const { metrics, flags } = summarizeSkillTelemetry(proc, activations, policy, now);
+      if (proc.skillPath) refreshGeneratedSkillLifecycleState(db, skillName, policy, now);
+      const procFresh = findGeneratedSkillProcedure(db, skillName) ?? proc;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/generated-skills.ts]: Fixed in becd6a2704a3 — buildGeneratedSkillTelemetryReport now uses refreshGeneratedSkillLifecycleState's returned ProcedureEntry and avoids the immediate second findGeneratedSkillProcedure scan.

File: extensions/memory-hybrid/services/procedure-promotion-policy.ts [hunk 5]
```diff
@@ -1089,12 +1092,39 @@ function hasValidationCheck(recipe: unknown, _task: string): boolean {
     return explicitValidationPattern.test(`${fields}\n${argFields}`);
   });
 }
+const skillMdDuplicateDigestCache = new Map<string, { mtimeMs: number; lower: string }>();
+
+/** Clears the SKILL.md mtime cache used by duplicate-skill detection (tests / --no-skill-duplicate-cache). */
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/procedure-promotion-policy.ts]: Fixed in becd6a2704a3 — the duplicate-skill cache-clear docstring now names the actual --bypass-skill-duplicate-cache flag, and the helper remains exported for tests/direct callers.

File: extensions/memory-hybrid/backends/migrations/facts-migrations.ts [hunk 6]
```diff
@@ -427,6 +427,83 @@ function migrateGeneratedSkillTelemetryTable(db: DatabaseSync): void {
...
+           gst_outcome_partial = (SELECT COUNT(*) FROM generated_skill_telemetry t WHERE t.procedure_id = p.id AND t.decision = 'selected' AND t.task_outcome = 'partial'),
+           gst_outcome_unknown = (SELECT COUNT(*) FROM generated_skill_telemetry t WHERE t.procedure_id = p.id AND t.decision = 'selected' AND (t.task_outcome IS NULL OR t.task_outcome = '' OR t.task_outcome NOT IN ('success','failure','partial'))),
+           gst_success_clear_total = (SELECT COUNT(*) FROM generated_skill_telemetry t WHERE t.procedure_id = p.id AND t.decision = 'selected' AND t.task_outcome = 'success' AND t.user_correction = 0),
+           gst_considered_total = (SELECT COUNT(*) FROM generated_skill_telemetry t WHERE t.procedure_id = p.id AND t.decision = 'considered'),
+           gst_skipped_total = (SELECT COUNT(*) FROM generated_skill_telemetry t WHERE t.procedure_id = p.id AND t.decision = 'skipped'),
+           gst_saved_tool_calls_sum = (SELECT COALESCE(SUM(t.saved_tool_calls), 0) FROM generated_skill_telemetry t WHERE t.procedure_id = p.id),
+           gst_saved_time_ms_sum = (SELECT COALESCE(SUM(t.saved_time_ms), 0) FROM generated_skill_telemetry t WHERE t.procedure_id = p.id),
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/migrations/facts-migrations.ts]: Fixed in becd6a2704a3 — migration backfill now clamps saved_tool_calls and saved_time_ms with the same positive-only CASE expressions as delta/rebuild; added a regression test for rebuild parity with negative saved values.

File: extensions/memory-hybrid/backends/workflow-store.ts [hunk 7]
```diff
@@ -416,15 +462,18 @@ export class WorkflowStore extends BaseSqliteStore {
...
+      const out = filtered.slice(0, resultLimit);
+      this.workflowPatternsMemo = {
+        revision: this.workflowPatternsRevision,
+        traceSampleLimit,
+        similarityThreshold: threshold,
+        minSuccessRate: minRate,
+        resultLimit,
+        patterns: out,
+      };
+      return out;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/workflow-store.ts]: Fixed in 842365989f547cf08a63614a3d5e667572baa2f3 — getPatterns now stores and returns defensive copies of memoized WorkflowPattern arrays, including nested toolSequence/exampleGoals arrays, and adds a regression test for caller mutation.

**Comment by markus-lassfolk** [comment]
@cursoragent review

Stewardship advance: PR #1451 ready for Council review at current head `47551faf229829be3d62c131ac99aa6796aba6d8`.

**Status at current head:**
- All review threads resolved ✓
- CI green, mergeable, UNSTABLE ✓
- No current-head Council
- Labels: `extensions size/XL stage/review` (no `fixes-needed`)

**PR title:** perf(memory-hybrid): workflow patterns telemetry and duplicate detection

**Council criteria:** Approve if — telemetry collection is correct, duplicate detection is accurate, no performance impact, UNSTABLE CI status understood and safe to merge.

`next_owner=Council` | `stage=council` | Please review and post verdict.

**Comment by markus-lassfolk** [comment]
PR #1451 stage/approved advance.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1451
Head SHA: 47551faf229829be3d62c131ac99aa6796aba6d8
stage: stage/approved (was stage/council)
blockers: none
action_taken: label advance stage/council → stage/approved
next_owner: merge gate; recheck after checks settle
proof: https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1451
